### PR TITLE
Emerald Warden Bot

### PR DESCRIPTION
**Pull request by Sparks1992**
Code for Emerald Warden hero.

### CODE REVIEW

File: EmeraldWarden_main.lua [hunk 1]
```diff
@@ -0,0 +1,342 @@
...
+}
+function object:SkillBuild()
+
+	local unitSelf = self.core.unitSelf
+
+	if not bSkillsValid then
+		skills.abilSilence = unitSelf:GetAbility(0)   -- 1st Skill -> Silencing Shot
+		skills.abilWolves = unitSelf:GetAbility(1)  -- 2nd Skill -> Hunter's Command
+		skills.abilOvergrowth = unitSelf:GetAbility(2)  -- 3rd Skill -> Overgrowth
+		skills.abilGawain = unitSelf:GetAbility(3)  -- Ultimate -> Summon Gawain
```
Comment (Anakonda) [inline EmeraldWarden_main.lua:97]: Gawain have 3 abilities that are in slots 5, 6 and 7. You still level up the ability at slot 3.

File: EmeraldWarden_main.lua [hunk 2]
```diff
@@ -0,0 +1,342 @@
...
+-----------------------
+
+-- Escape combo for running away
+
+function behaviorLib.CustomRetreatExecute(botBrain)
+	bActionTaken = false
+
+	local unitSelf = core.unitSelf
+	local unitSelfPosition = unitSelf:GetPosition()
+	local bCanSee = unitTarget and core.CanSeeUnit(botBrain, unitTarget)
```
Comment (Anakonda) [inline EmeraldWarden_main.lua:297]: unitTarget is nil here. You need to pick target from core.localUnits["EnemyHeroes"]

**Review by Anakonda — CHANGES_REQUESTED** [review]

File: EmeraldWarden_main.lua (hunk 1 above)
Comment (Sparks1992) [reply EmeraldWarden_main.lua:97]: and what should i level up?... gawain has 1 ability at lvl 1, 2 abilities at lvl 2 and 3 at lvl 3

File: EmeraldWarden_main.lua (hunk 2 above)
Comment (Sparks1992) [reply EmeraldWarden_main.lua:297]: i changed unitTarget to core.localUnits["EnemyHeroes"]

File: EmeraldWarden_main.lua (hunk 1 above)
Comment (Sparks1992) [reply EmeraldWarden_main.lua:97]: should i add skills.abilGawain = unitSelf:GetAbility(5)
skills.abilGawain = unitSelf:GetAbility(6)
skills.abilGawain = unitSelf:GetAbility(7)
and at lvl 6, lvl-up 3-rd, at 11, 5th and at 16 6th?

File: EmeraldWarden_main.lua [hunk 3]
```diff
@@ -0,0 +1,342 @@
...
+-----------------------
+
+-- Escape combo for running away
+
+function behaviorLib.CustomRetreatExecute(botBrain)
+	bActionTaken = false
+
+	local unitSelf = core.unitSelf
+	local unitSelfPosition = unitSelf:GetPosition()
+	local bCanSee = core.localUnits["EnemyHeroes"] and core.CanSeeUnit(botBrain, unitTarget)
```
Comment (Anakonda) [inline EmeraldWarden_main.lua]: This is list of units. You need to pick one of them and set it to unitTarget. Preferably one that is in range.

File: EmeraldWarden_main.lua (hunk 1 above)
Comment (Anakonda) [reply EmeraldWarden_main.lua:97]:
```
skills.abilStrike = unitSelf:GetAbility(5)
skills.abilHeal = unitSelf:GetAbility(6)
skills.abilStorm = unitSelf:GetAbility(7)
```
Only add the ones you use later. Also note that as they are passives you cant activate so `:CanActivate()` is always false, use `:GetActualRemainingCooldownTime()` and check if it is 0

**Review by Anakonda — CHANGES_REQUESTED** [review]

File: EmeraldWarden_main.lua (hunk 1 above)
Comment (Sparks1992) [reply EmeraldWarden_main.lua:97]: they're all passive so i don't have to use them... i should only add the threshold for them for increased aggressiveness?

**Comment by Sparks1992** [comment]
Please review the changes for skills and threshold

File: EmeraldWarden_main.lua (hunk 3 above)
Comment (Sparks1992) [reply EmeraldWarden_main.lua]: would it be ok if i do it like this?
	local unitTarget = behaviorLib.heroTarget
	local bCanSee = unitTarget and core.CanSeeUnit(botBrain, unitTarget)
	if unitTarget == nil or not unitTarget:IsValid() then
		return false --can not execute, move on to the next behavior
	end

**Pull request closed without merging**